src/util/mpir_hwtopo.c

```diff
@@ -11,53 +11,531 @@
+                if (io_device->attr->osdev.type == HWLOC_OBJ_OSDEV_GPU) {
+                    if ((*(name + strlen("gpu")) != '\0') &&
+                        atoi(name + strlen("gpu")) != io_device->logical_index) {
+                        continue;
+                    }
+                } else {
+                    continue;
+                }
+            }
+
+            non_io_ancestor = hwloc_get_non_io_ancestor_obj(hwtopo.hwloc_topology, io_device);
+            while (!hwloc_obj_type_is_normal(non_io_ancestor->type))
+                non_io_ancestor = non_io_ancestor->parent;
+            MPIR_Assert(non_io_ancestor && non_io_ancestor->depth >= 0);
+
+            if (!hwloc_obj_is_in_subtree
+                (hwtopo.hwloc_topology, obj_containing_cpuset, non_io_ancestor))
+                continue;
+
+            break;
+        }
+
+        if (non_io_ancestor) {
+            MPIR_hwtopo_class_e class = get_type_class(non_io_ancestor->type);
+            gid = HWTOPO_GET_GID(class, non_io_ancestor->depth, non_io_ancestor->logical_index);
+        }
+
+        return gid;
+    }
+#endif
+
+    MPIR_hwtopo_type_e type = MPIR_hwtopo_get_type_id(name);
+    if (type != MPIR_HWTOPO_TYPE__MAX)
+        gid = MPIR_hwtopo_get_obj_by_type(type);
+    else
+        MPID_Get_node_id(MPIR_Process.comm_world, MPIR_Process.comm_world->rank, &gid);
+
+    return gid;
 }
 
 int MPIR_hwtopo_mem_bind_single(void *baseaddr, size_t len, MPIR_hwtopo_gid_t gid)
 {
+    int ret = MPI_SUCCESS;
+
+#ifdef HAVE_HWLOC
+    const struct hwloc_topology_support *support =
+        hwloc_topology_get_support(hwtopo.hwloc_topology);
+    if (!support->membind->set_area_membind) {
+        fprintf(stderr,
+                "%s: hwloc_set_area_membind() is not supported, skipping memory binding\n",
+                __func__);
+        return MPI_ERR_OTHER;
+    }
+
+    hwloc_membind_policy_t policy = HWLOC_MEMBIND_BIND;
+    hwloc_membind_flags_t flags = HWLOC_MEMBIND_STRICT;
+
+    int hwloc_obj_index = HWTOPO_GET_INDEX(gid);
+    int hwloc_obj_depth = HWTOPO_GET_DEPTH(gid);
+
+    hwloc_obj_t hwloc_obj =
+        hwloc_get_obj_by_depth(hwtopo.hwloc_topology, hwloc_obj_depth, hwloc_obj_index);
+
+    hwloc_bitmap_t bitmap = hwloc_bitmap_alloc();
+    hwloc_bitmap_or(bitmap, bitmap, hwloc_obj->nodeset);
+
+    if (hwloc_obj->type == HWLOC_OBJ_NUMANODE) {
+        flags |= HWLOC_MEMBIND_BYNODESET;
+    } else {
+        fprintf(stderr, "%s: object type not valid, skipping memory binding\n", __func__);
+        return MPI_ERR_OTHER;
+    }
+
+    ret = hwloc_set_area_membind(hwtopo.hwloc_topology, baseaddr, len, bitmap, policy, flags);
+
+    hwloc_bitmap_free(bitmap);
+#endif
+
+    return ret;
 }
 
 int MPIR_hwtopo_mem_bind_multi(int size, void **baseaddr, size_t * len, MPIR_hwtopo_gid_t * gid)
 {
+    int ret = MPI_SUCCESS;
+
+    for (int i = 0; i < size; i++)
+        ret = MPIR_hwtopo_mem_bind_single(baseaddr[i], len[i], gid[i]);
+
+    return ret;
 }
 
```

Comment:
I thought this function is to bind a single block of memory to multiple `gid` with `INTERLEAVE` policy. If it is just individual `MPIR_hwtopo_mem_bind_single`, then we don't need such function. The "user" can directly do the `for-loop` and it is easier to read than call an opaque function.

Reply:
No that would be another function. But since we don't use it I haven't added it. That would be `MPIR_hwtopo_interleave_mem()`

Reply:
I see. In that case, I would suggest remove this function.